server/gt_test.py

```diff
@@ -0,0 +1,199 @@
+    for key in keys:
+        del filters[key]
+    fout = "%s_%s_%s_nonequalGT.tsv" % (variant, '_'.join(f'{key}_{value}' for key, value in filters.items()), gt_download)
+
+    # get non-equal rows
+    ind = dfs[0] != dfs[1]
+    tot = dfs[2][list(ind['gt'])]
+    tot['vcf_gt'] = list(dfs[0].loc[list(ind['gt'])]['gt'])
+    tot = tot.reset_index(drop=True)
+
+    # save output if specified so
+    tot.to_csv(fout, sep="\t", index=False)
+    print("Saved to", fout)
+
+
+def assert_equal_df(dfs):    
+    # assert data frames are equal
+    try:
+        assert_frame_equal(dfs[0], dfs[1], check_dtype=False)
+    except AssertionError as e:
+        print("ERROR :: %s." % e)
+        return False
+    else:
+        return True
+
+
+def getData(variant, filters, data_type):
+    print("\nGetting genotypes for variant %s from %s data." % (variant, data_type))
+    chr, pos, ref, alt = variant.split('-')
+    if chr.startswith('chr'):
+        chr = re.sub(r"[^0-9]+", "", chr)
+    if chr == 'X' or chr == '23':
+        chr_search = 'X'
+        chr = '23'
+    else:
+        chr_search = chr
+    chr = int(chr)
+    pos = int(pos)
+    with app.app_context():
+        # get write data from genotype browser
+        try:
+            response = fetch.write_variants(variant, filters.copy(), data_type)
+        except ValueError as e:
+            # print("ValueError :: %s" % e)
+            dat_gb_full = None
+        else:
+            text = response.data
+            dat_gb_full = pd.read_csv(io.StringIO(text.decode()), sep='\t', low_memory=False)
+            # dat_gb = dat_gb[['FINNGENID', 'gt']]
+            dat_gb_full.index = list(dat_gb_full['FINNGENID'])
+        
+        # extract genotypes directly from the vcf directly
+        if data_type == 'imputed':
+            vcf = config['vcf_files']['imputed_data'][chr-1]
+        else:
+            vcf = config['vcf_files']['chip_data'][chr-1]
+        
+        # check vcf files
+        x = pysam.TabixFile(vcf, parser=None)
+        gt_data = x.fetch("chr%s" % chr_search, pos - 1, pos)
+        gt = None
+        for row in gt_data:
+            data = row.split('\t')      
+            if data[3] == ref and data[4] == alt:
+                gt = data
+                break
+        
+        # construct data frame
+        if gt is None:
+            dat_vcf = None
+        else:
+            samples = x.header[len(x.header) - 1].split('\t')
+            d = pd.DataFrame({'FINNGENID': samples, 'gt_': gt})
+            d = d.iloc[9:, ]
+            d['gt'] = [item.split(':')[0] for item in d['gt_']]
+            dat_vcf = d[['FINNGENID', 'gt']]
+            dat_vcf.index = list(dat_vcf['FINNGENID'])
+        
+        # re-order
+        dat_gb = None
+        if dat_gb_full is not None:
+            dat_gb = dat_gb_full[['FINNGENID', 'gt']].copy()
+            dat_vcf = dat_vcf.loc[list(dat_gb.index), :]
```

Comment:
dat_vcf can be None at this point if the variant is not found in the VCF and this would then break?

Reply:
Thank you for your comments! That would indeed break the code. I made the updates accordingly, please take a look when you have time.